src/_pytest/python.py

```diff
@@ -1000,49 +1023,59 @@ def _resolve_arg_ids(
         :rtype: List[str]
         :return: the list of ids for each argname given
         """
-        idfn = None
-        if callable(ids):
+        if ids is None:
+            idfn = None
+            ids_ = None
+        elif callable(ids):
             idfn = ids
-            ids = None
-        if ids:
-            func_name = self.function.__name__
-            ids = self._validate_ids(ids, parameters, func_name)
-        ids = idmaker(argnames, parameters, idfn, ids, self.config, item=item)
-        return ids
-
-    def _validate_ids(self, ids, parameters, func_name):
+            ids_ = None
+        elif not ids:
+            # Special case: https://github.com/pytest-dev/pytest/issues/1849
```

Comment:
Might get handled also by _validate_ids?  (handling empty list/returing there).

It looks like a lot of "validation" happens before it already (also before, but a bit more now?).
Not too bad, but just a thought.

Reply:
The following works:

```diff
diff --git a/src/_pytest/python.py b/src/_pytest/python.py
index e4d7cbf86..802298641 100644
--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -1029,10 +1029,6 @@ class Metafunc:
         elif callable(ids):
             idfn = ids
             ids_ = None
-        elif not ids:
-            # Special case: https://github.com/pytest-dev/pytest/issues/1849
-            idfn = None
-            ids_ = []
         else:
             idfn = None
             ids_ = self._validate_ids(ids, parameters, self.function.__name__)
@@ -1053,7 +1049,8 @@ class Metafunc:
                 raise TypeError("ids must be a callable or an iterable")
             num_ids = len(parameters)
 
-        if num_ids != len(parameters):
+        # num_ids == 0 is a special case: https://github.com/pytest-dev/pytest/issues/1849
+        if num_ids != len(parameters) and num_ids != 0:
             msg = "In {}: {} parameter sets specified, with different number of ids: {}"
             fail(msg.format(func_name, len(parameters), num_ids), pytrace=False)
```

If you think it's better (I do!) I can include it.

Reply:
Yes, I also think it is better.